scanner/decisions/0010-nvd-cvss-scores.md

```diff
@@ -0,0 +1,192 @@
+In reality, the CVSS scores that are shown are from NVD except in the following cases:
+
+* RHEL-base images
+  * Red Hat-provided CVSS scores/severities are used.
+* OSV-provided data
+  * If OSV.dev has a CVSS score, then that score is used.
+    * The severity is derived from the score based on the CVSS version's specification.
+  * Note: From personal experience, it is very common for the score displayed by OSV is the same as NVD.
+
+Currently, there is only a need to have CVSS scores from the vendor (if available) and NVD (if available), no more.
+However, it is possible this requirement changes in the future, and we may want to support more than just two
+sources for vulnerability data.
+
+The focus is on CVEs and RHSAs, RHEAs, and RHBAs. This document will not consider other types of advisories such as
+Ubuntu's USNs.
+
+The current API looks like the following:
+
+```
+message VulnerabilityReport {
+  message Vulnerability {
+    enum Severity {
+      SEVERITY_UNSPECIFIED = 0;
+      SEVERITY_LOW = 1;
+      SEVERITY_MODERATE = 2;
+      SEVERITY_IMPORTANT = 3;
+      SEVERITY_CRITICAL = 4;
+    }
+    message CVSS {
+      message V2 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      message V3 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      V2 v2 = 1;
+      V3 v3 = 2;
+    }
+    ...
+    string severity = 6;
+    ...
+    CVSS cvss = 12;
+  }
+  ...
+}
+```
+
+## Decision
+
+The VulnerabilityReport API will be extended as follows:
+
+```
+message VulnerabilityReport {
+  message Vulnerability {
+    enum Severity {
+      SEVERITY_UNSPECIFIED = 0;
+      SEVERITY_LOW = 1;
+      SEVERITY_MODERATE = 2;
+      SEVERITY_IMPORTANT = 3;
+      SEVERITY_CRITICAL = 4;
+    }
+    message CVSS {
+      enum Source { <-- New field
+        SOURCE_UNKNOWN = 0;
+        SOURCE_RED_HAT = 1;
+        SOURCE_OSV = 2;
+        SOURCE_NVD = 3;
+      }
+      message V2 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      message V3 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      V2 v2 = 1;
+      V3 v3 = 2;
+      Source source = 3; <-- New field
+      string url = 4;    <-- New field
+    }
+    ...
+    string severity = 6;
+    ...
+    CVSS cvss = 12;
+    repeated CVSS cvss_metrics = 13; <-- New field.
+  }
+  ...
+}
+```
+
+There will be a new type plus two new fields added:
+
+* **enum** `CVSS.Source`
+  * This is the **enum** that specifies the supported CVSS metrics data sources.
+* **field** `CVSS.source`
+  * This **field** specifies the source of the particular CVSS metrics.
+* `CVSS.url`
+  * This specifies the CVSS score source URL, aiding both API and UI users in tracking the origin of this metric.
+* `cvss_metrics`
+  * This is a list of each unique CVSS metric based on the source.
+
+The original `cvss` field will remain and will continue to represent the Scanner's preferred CVSS score.
```

Comment:
Other options that also allow providing a "preferred score", examples:

1. A flag in the CVSS message `isPreferred` or similar.
2. A new field to specify the CVSS source that is preferred (by `Source`).

Questions:

1. Were you planning to show the preferred score in `cvss` and all scores (including the preferred) in `cvss_metrics`?

2. Why one additional field was selected? This sounds more suitable for keeping backward compatibility, but since Central and Scanner are deployed together, and the protobufs are not stored, I don't think keeping the `cvss` field populated is necessary?

3. Why do we need to indicate which is preferred? If you think about it, Scanner falls back to NVD when the security source is not available, and Central could easily make the same decision, or any other decision, about which CVSS score to show. Can you help me understand the value of Scanner providing a "preferred" CVSS score?

Reply:
q1: Currently the scope is to make sure the clients can access NVD CVSS score. The UX had a plan to show all scores but that's beyond 4.6
q2: It is for the purpose of keeping backward compatibility, I am not sure if we want to break that. @RTann what do you think?
q3: I wanted to keep the original `cvss` field mainly for the backward compatibility reason

Reply:
> Currently the scope is to make sure the clients can access NVD CVSS score. The UX had a plan to show all scores but that's beyond 4.6

The plan makes sense. I wanted to understand if you were planning to have the field `cvss` point to "preferred score", but then `cvss_metrics` will be a list of all scores, including the one pointed out by `cvss`?

Example:

```
{"cvss": "{ ..., "source": "REDHAT", ...}",
 "cvss_metrics": [{ ..., "source": "REDHAT", ...}, { ..., "source": "NVD", ... }]
}
```

Or:

```
{"cvss": "{ ..., "source": "REDHAT", ...}",
 "cvss_metrics": [{ ..., "source": "NVD", ... }]
}
```

> It is for the purpose of keeping backward compatibility, I am not sure if we want to break that. @RTann what do you think?
>
> ...
> 
> > I wanted to keep the original cvss field mainly for the backward compatibility reason

Understood. I don't think backward compatibility here is relevant because we don't store this proto file anywhere (to my knowledge) and there are no forward-compatible readers. Meaning, Central and Scanner will be deployed together, and both will understand the new proto; If my understanding is true, then we can just ignore this field from now on (mark it as deprecated, etc), and we can provide all CVSS metrics in the new "cvss_metrics". Central (or any consumer) can decide which source to use; We don't need to specify which one is "preferred" because that decision is not Scanner's to make.

In some places, like E2E tests and maybe UTs, you would have to account for the fact that "cvss" is not not being populated; That means is additional work to update these parts. As an option, maybe we can populate the "cvss" with what we do today but mark the field as deprecated and actually move consumers to the new "cvss_metrics" and then stop populating the field.

I prefer moving away from "cvss" now.

Reply:
I have marked the `cvss` and `link` fields as deprecated and have aligned with other stakeholders in the NVD project, who are supportive of this decision. The core workflow team is particularly aware of the changes, and we all agree to make the logic less dependent on these deprecated fields.